### ISSUE TITLE
Swipe up quickly and then swipe down quickly, the modal won't dismiss?

### ISSUE THREAD

**Issue by lchenfox** [HeathWang/HWPanModal#120]
**Describe the bug**

Firstly, I swipe up the modal view quickly, and then swipe down quickly, the modal view won't dismiss as expected. It should dismiss when I swipe down quickly. In addition, it can dismiss if I swipe down quickly without the swiping up first.

**To Reproduce**
1. Swipe up quickly first.
2. Swipe down quickly.
3. The modal does not dismiss as expected.

**Expected behavior**

The modal should always dismiss as long as I swipe down quickly each time.

**Screenshots**
No screenshots needed.

**Smartphone (please complete the following information):**
 - Device: `iPhone7`
 - OS: `14.8.1 `
 - Version `HWPanModal (= 0.8.9)`

**Additional context**

I found some code snippets here.

```
- (void)handleDragUpState:(PresentationState)state {
    switch (state) {
        case PresentationStateLong:
            [self transitionToState:PresentationStateLong];
            [self cancelInteractiveTransition];
            break;
        case PresentationStateMedium:
            [self transitionToState:PresentationStateLong];
            [self cancelInteractiveTransition];
            break;
        case PresentationStateShort:
            [self transitionToState:PresentationStateMedium];
            [self cancelInteractiveTransition];
            break;
        default:
            break;
    }
}
```

if I swipe up quickly first, the `case PresentationStateShort:` will be called and the `state` will be `PresentationStateMedium`. Then I swipe down quickly to dismiss the modal, but It does not because the `state` has been reset to `PresentationStateMedium`.

Now as a temporary solution, I have to force to reset the `state` via overriding the super method `- (void)didChangeTransitionToState:(PresentationState)state` in `HomePanModalSubView`(Inherit from `HWPanModalContentView`) as follows:

```
- (void)didChangeTransitionToState:(PresentationState)state {
    if (state != PresentationStateShort) {
         // Force to reset the state to PresentationStateShort] here.
        [self hw_panModalTransitionTo:PresentationStateShort];
    } 
}
```

Is it a bug or is that right for the solution above? Or it's designed as expected? Any answer will be appreciated.

**Comment by HeathWang** [comment]
it seems you set the mode as short -> medium -> long, now the design like this.
If you just set short -> long mode, it should not have this issue.

**Comment by lchenfox** [comment]
> I check the drag logic, the design is: When the short, medium, long are different, it will go to next state, unless you are in short, you can drag to dismiss directly.

Thank you for reply. If that, swiping up first(the `state` changed to `PresentationStateMedium`), then I have to swipe down twice to dismiss the modal, which may be redundant. 

As the solution seen above, can I do like that to insure that the modal can always swipe down to dismiss directly each time even after firstly swiping up?

**Comment by lchenfox** [comment]
@HeathWang 

Now I try another way to avoid this problem like 

```
// Override super method.  Returns NO.
- (BOOL)shouldTransitionToState:(PresentationState)state {
    return  NO;
}
```

It also seems to solve the problem. As for two ways, which  one is better?

**Comment by HeathWang** [comment]
yes, if you have tested, you can do like that.

**Comment by lchenfox** [comment]
Ok, thanks a lot. Closing the issue.

**Comment by HeathWang** [comment]
I have published HWPanModal (0.9.2).
Add a new API `- (void)hw_dismissAnimated:(BOOL)animated completion:(void (^)(void))completion`.

now for your case, you can handle like this:

```
// override this protocol
- (void)didEndRespondToPanModalGestureRecognizer:(UIPanGestureRecognizer *)panGestureRecognizer {
    CGPoint velocity = [panGestureRecognizer velocityInView:self.hw_contentView];
     // decide the velocity.y value you want to dismiss
     // for ViewController, you can call `isBeingDismissed`, for view, you can get the dismiss start from `- (void)panModalWillDismiss;` delegate
    if (velocity.y > 200 && !self.isBeingDismissed) {
        [self hw_dismissAnimated:YES completion:^{
            // do something.
        }];
    }
}
```
